File: books-practice/Data Structures and  Algorithms in Java/src/chap_6/stack/java/arraystack/ArrayStack.java
package arraystack;
public class  ArrayStack<E> implements Stack<E> {
  public static final int CAPACITY = 1000; // default array capacity
  private E[] S; // generic array used for storage
  private int top = -1; // index of the top element in stack

  public ArrayStack() {
    this(CAPACITY);
  }

  @SuppressWarnings({ "unchecked" })
  public ArrayStack(int capacity) {
    S = (E[]) new Object[capacity]; // safe cast; compiler may give warning
  }

  @Override
  public int size() {
    return (top + 1);
  }

  @Override
  public boolean isEmpty() {
    return (top == -1);
  }

  /**
   * `top = top + 1;
   * S[top] = e;`
   */
  @Override
  public void push(E e) throws IllegalStateException {
    if (size() == S.length)
      throw new IllegalStateException("Stack is full");
    S[++top] = e; // increment top and then store new item
  }

  @Override
  public E top() {
    if (isEmpty())
      return null;
    return S[top];
  }

  @Override
  public E pop() {
    if (isEmpty())
      return null;
    E answer = S[top];
    S[top] = null; // dereference to help garbage collection
    top--;
    return answer;
  }
}

File: books-practice/Data Structures and  Algorithms in Java/src/chap_6/stack/typescript/arraystack/arraystack.ts
export interface Stack<E> {
  size(): number;
  isEmpty(): boolean;
  push(e: E): void;
  top(): E | undefined;
  pop(): E | undefined;
}

export class ArrayStack<E> implements Stack<E> {
  private CAPARITY: number = 10;
  private S: (E | undefined)[];
  private toppest: number = -1;

  constructor() {
    this.S = new Array(this.CAPARITY);
  }

  size(): number {
    return this.toppest + 1;
  }
  isEmpty(): boolean {
    return this.toppest === -1;
  }
  push(e: E): void {
    if (this.size() == this.S.length) {
      throw Error('Stack if full');
    }
    this.toppest++;
    this.S[this.toppest] = e;
  }

  top(): E | undefined {
    if (this.isEmpty()) undefined;
    return this.S[this.toppest];
  }
  pop(): E | undefined {
    if (this.isEmpty()) undefined;
    let r = this.S[this.toppest];
    this.S[this.toppest] = undefined;
    this.toppest--;
    return r;
  }
  loop(): void {
    for (let i = 0; i < this.S.length; i++) {
      console.log(this.S[i]);
    }
  }
}

const testcase = new ArrayStack<Number>();
testcase.push(1);
testcase.push(10);
testcase.push(10);
testcase.push(10);
testcase.pop();

console.log(testcase.loop())